### ISSUE TITLE
Add Tokens for Import and Export Names in JS? (Local, Import and Export Names)

### ISSUE THREAD

**Issue by karlhorky** [PrismJS/prism#2528]
**Information**
- Language: JavaScript
- Plugins: I tried with JS-Extras too

Does the problem still occur in the latest version of Prism? Yes

**Description**
Would it be possible to add tokens for the import and export names in JavaScript? (see patterns below taken from the [import examples](https://tc39.es/ecma262/#table-40) and [export examples](https://tc39.es/ecma262/#table-42) in the ECMAScript spec)

See below - currently they are not receiving any tokenization. To be clear, I'm referring to the local names, import names and export names with names such as `v`, `x`, `v2` and `ns`:

![Screen Shot 2020-08-26 at 10 48 37](https://user-images.githubusercontent.com/1935696/91283902-80776480-e78b-11ea-8e47-663692d2d000.png)

![Screen Shot 2020-08-26 at 11 00 05](https://user-images.githubusercontent.com/1935696/91283921-85d4af00-e78b-11ea-81aa-8d827c78b7a4.png)

I also tried using the `js-extras` language as follows, but that didn't seem to help:

https://codesandbox.io/s/prismjs-forked-wu91d

**Code snippet**

The code being highlighted incorrectly:

```js
import v from "mod";
import {x as v2} from "mod";
import * as ns from "mod";

export * as ns from "mod";
export {v as x} from "mod";
export {x} from "mod";
export {v as x};
export {x};
```

**Comment by RunDevelopment** [comment]
What's the use case for this? 

Also, we usually highlight concepts, so what is the general idea here?

Syntax-wise, from the [`import`](https://tc39.es/ecma262/#sec-imports)s you want all identifiers (e.g. `import { x as y } from "mod"` both `x` and `y`) and from [`exports`](https://tc39.es/ecma262/#sec-exports), you want the same, so no export declarations (e.g. `export const foo;`, `export default function foo(){}`, and `export class Foo {}`), right?

**Comment by karlhorky** [comment]
Use case is being able to be more specific about import names for themes like this:

https://github.com/PrismJS/prism-themes/pull/103

---

> Also, we usually highlight concepts, so what is the general idea here?

Right ok, I suppose the overarching concept is actually "all identifiers". In addition to the variables in the original post, I would think that `a`, `b`, and `foo` would receive tokens eg. `.token.identifier` or something:

<img width="276" alt="Screen Shot 2020-08-27 at 11 54 59" src="https://user-images.githubusercontent.com/1935696/91425998-2db9ad80-e85c-11ea-9c39-312574bef9be.png">

As for a concrete use case about that (also irrespective of the `import` and `export` rules), the VS Code dark theme currently has a mismatch between the identifier colors here:

![Screen Shot 2020-08-27 at 12 03 48](https://user-images.githubusercontent.com/1935696/91426922-745bd780-e85d-11ea-9a88-e11602efec66.png)

Maybe for import and export names they could be subclassed as `.token.import .token.identifier` or `.token.identifier.import-identifier` or similar, in case the styling changes between import identifiers and regular identifiers in a theme.

---

If it makes sense, I can change this issue to be about "identifiers" in general.

**Comment by karlhorky** [comment]
I suppose another consideration about assigning tokens to identifiers is if it's possible + feasible to handle the other locations where an identifier is used (eg. `c`, `d`, `e`, `f`, `g`, `h`):


![Screen Shot 2020-08-27 at 12 31 11](https://user-images.githubusercontent.com/1935696/91431883-48daec00-e861-11ea-9338-6ea645565c57.png)

**Comment by RunDevelopment** [comment]
> [...] the VS Code dark theme currently has a mismatch between the identifier colors here:

Maybe a bit on how VSC highlights variables in JS/TS. VSC uses the TS language server to parse and analyze JS and TS files (among other things). A while ago they introduced [sematic highlighting](https://code.visualstudio.com/api/language-extensions/semantic-highlight-guide). This means, that variables are not highlighted by syntax alone but with a deeper understanding of the type and other properties of the variable (powered by the TS language server). 

This is why VSC can highlight `bar` as a function even though, syntactically speaking, it's just another variable.
![image](https://user-images.githubusercontent.com/20878432/91445138-08399d80-e876-11ea-849d-7cc4e4d128e3.png)

VSC can do this not only for the type of a variable (class/function/other) but also for its mutability. Readonly variables/properties will be highlighted with a darker shade of blue:
![image](https://user-images.githubusercontent.com/20878432/91445601-9f065a00-e876-11ea-89b8-95ed5f63a279.png)

This is incredibly useful because it gives a lot of information to the developer that is not obvious from syntax alone.

However, that's the problem for us. We need more information than just syntax to handle the nested properties correctly. For flat variables, syntax alone is technically enough, but...

> if it's possible + feasible to handle the other locations where an identifier is used [...].

No. Tracking variables requires an understanding of scope. This is hard enough with an AST but we only have a token stream to work with. This is one of the things that Prism simply can't do.

**Comment by karlhorky** [comment]
Right yeah, I understand that we won't get to 100% of the capabilities of an editor with type information.

But for the following, maybe regexes / config like the following would be possible:

```js
const a = 1;
let b = 2,
  c = 3;

export const foo = 1;
```

**Config:**

```
	'declaration': {
		pattern: /(const|let|var)\s+([\n\s]*[_$a-zA-Z\xA0-\uFFFF][$\w\xA0-\uFFFF]*([\n\s]*=[\n\s]*[^,;]+)?[\n\s]*[,;])+/,
		inside: {
			'identifier': {
				pattern: /\b((?!const|let|var)[_$a-zA-Z\xA0-\uFFFF][$\w\xA0-\uFFFF]*)\b/,
				alias: 'string'
			},
			rest: Prism.languages.javascript
        }
    }
```


For the more complex examples, maybe the contents inside the parentheses of function calls could be parsed with a similar pattern to `identifier` above.

Technically, it's an `expression` - maybe a regular expression could be developed for that too... 🤔

```js
const d = e * f;

console.log(g.z);

otherFn(h * i);
```

**Comment by RunDevelopment** [comment]
This will be hard because we have to deal with [JS' semicolon insertion](http://www.bradoncode.com/blog/2015/08/26/javascript-semi-colon-insertion/). Basically, we don't know where a variable declaration ends. To get this right, we'd have to match a single valid JS expression. The formal language of a single JS expression is context-free and cannot be expressed by a regex. We can usually work around this because we usually know the next character after the expression (e.g. a comma or closing bracket).

If we assume semicolons, it's doable but otherwise, ... I doubt it.

We could settle for the first declaration. That's easy and, hopefully, nobody uses these merged variable declarations.

However, even then: Will this be beneficial?
The end goal is to give const variables a different color to more closely resemble VSC's highlighting. But I'd argue that this would hurt readability because const variables will appear differently when used compared to when they're declared.

If I remember correctly, VSC used to highlight all variable declarations in a different color but then dropped it again in the next release (I assume for similar reasons).

**Comment by karlhorky** [comment]
> The formal language of a single JS expression is context-free and cannot be expressed by a regex

> Will this be beneficial?

> const variables will appear differently when used compared to when they're declared

Ok, I can understand that this goes a bit far then. You're right that the benefit is small here and I agree that it would be confusing.

Maybe we stick with `import`s` and `export`s like this issue's description and forget about declarations for now?

```js
import v from "mod";
import {x as v2} from "mod";
import * as ns from "mod";

export * as ns from "mod";
export {v as x} from "mod";
export {x} from "mod";
export {v as x};
export {x};
```

**Comment by RunDevelopment** [comment]
Ok. Just one question before I make the pull: Should `as {name}` identifiers get a different class?

---

Also, the motivation for this feature is to more closely resemble VSC's theme, right? But that uses the normal variable color for imports/exports, so how do you intend to use the feature?

![image](https://user-images.githubusercontent.com/20878432/91466397-58712980-e88f-11ea-8920-6a8becfa6fd1.png)

**Comment by karlhorky** [comment]
Right so if you look at [the demo](https://objective-williams-02e373.netlify.app/), the `maybe-class-name` token is light green, whereas the imports are light blue:

![Screen Shot 2020-08-27 at 18 07 40](https://user-images.githubusercontent.com/1935696/91467126-6ecbb500-e890-11ea-8caa-658a20bbf4df.png)

![Screen Shot 2020-08-27 at 18 07 36](https://user-images.githubusercontent.com/1935696/91467140-7428ff80-e890-11ea-89a4-47e8e43d0c18.png)

**Comment by karlhorky** [comment]
I suppose an alternative that may work could be (`#9cdcfe` is the light blue color):

```css
.token.keyword.module + .token.maybe-class-name,
.token.keyword.module + .token.punctuation + .token.maybe-class-name {
    color: #9cdcfe;
}
```

**Comment by karlhorky** [comment]
> Should `as {name}` identifiers get a different class?

I guess the same token name should be fine?

**Comment by karlhorky** [comment]
Addressed in #2533 🙌 🚀

**Comment by RunDevelopment** [comment]
Closed by #2533.

**Comment by karlhorky** [comment]
Thanks @RunDevelopment !
